File: app/src/main/java/forpdateam/ru/forpda/presentation/topics/TopicsPresenter.java
package forpdateam.ru.forpda.presentation.topics;

import android.os.Bundle;

import com.arellomobile.mvp.InjectViewState;

import forpdateam.ru.forpda.api.reputation.Reputation;
import forpdateam.ru.forpda.api.reputation.models.RepData;
import forpdateam.ru.forpda.api.reputation.models.RepItem;
import forpdateam.ru.forpda.api.topcis.models.TopicItem;
import forpdateam.ru.forpda.api.topcis.models.TopicsData;
import forpdateam.ru.forpda.common.IntentHandler;
import forpdateam.ru.forpda.common.mvp.BasePresenter;
import forpdateam.ru.forpda.model.repository.reputation.ReputationRepository;
import forpdateam.ru.forpda.model.repository.topics.TopicsRepository;
import forpdateam.ru.forpda.presentation.reputation.ReputationView;
import forpdateam.ru.forpda.ui.TabManager;
import forpdateam.ru.forpda.ui.fragments.TabFragment;
import forpdateam.ru.forpda.ui.fragments.forum.ForumFragment;
import forpdateam.ru.forpda.ui.fragments.search.SearchFragment;
import io.reactivex.disposables.Disposable;

/**
 * Created by radiationx on 03.01.18.
 */

@InjectViewState
public class TopicsPresenter extends BasePresenter<TopicsView> {

    private int id = 0;
    private int currentSt = 0;
    private TopicsData currentData;

    private TopicsRepository topicsRepository;

    public TopicsPresenter(TopicsRepository topicsRepository) {
        this.topicsRepository = topicsRepository;
    }

    public void setId(int id) {
        this.id = id;
    }

    public int getId() {
        return id;
    }

    public void setCurrentData(TopicsData data) {
        currentData = data;
    }

    public TopicsData getCurrentData() {
        return currentData;
    }

    public void loadTopics() {
        Disposable disposable
                = topicsRepository.getTopics(id, currentSt)
                .doOnTerminate(() -> getViewState().setRefreshing(true))
                .doAfterTerminate(() -> getViewState().setRefreshing(false))
                .subscribe(repData1 -> getViewState().showTopics(repData1), this::handleErrorRx);

        addToDisposable(disposable);
    }

    public void loadPage(int st) {
        currentSt = st;
        loadTopics();
    }

    public void openForum() {
        Bundle args = new Bundle();
        args.putInt(ForumFragment.ARG_FORUM_ID, id);
        TabManager.get().add(ForumFragment.class, args);
    }

    public void openSearch() {
        String url = "https://4pda.ru/forum/index.php?act=search&source=all&forums%5B%5D=" + id;
        Bundle args = new Bundle();
        args.putString(TabFragment.ARG_TAB, url);
        TabManager.get().add(SearchFragment.class, args);
    }

    public void onItemClick(TopicItem item) {
        if (item.isAnnounce()) {
            Bundle args = new Bundle();
            args.putString(TabFragment.ARG_TITLE, item.getTitle());
            IntentHandler.handle(item.getAnnounceUrl(), args);
            return;
        }
        if (item.isForum()) {
            IntentHandler.handle("https://4pda.ru/forum/index.php?showforum=" + item.getId());
            return;
        }
        Bundle args = new Bundle();
        args.putString(TabFragment.ARG_TITLE, item.getTitle());
        IntentHandler.handle("https://4pda.ru/forum/index.php?showtopic=" + item.getId() + "&view=getnewpost", args);
    }

    public void onItemLongClick(TopicItem item) {
        getViewState().showItemDialogMenu(item);
    }
}

File: app/src/main/java/forpdateam/ru/forpda/presentation/topics/TopicsPresenter.kt
package forpdateam.ru.forpda.presentation.topics

import com.arellomobile.mvp.InjectViewState
import forpdateam.ru.forpda.common.mvp.BasePresenter
import forpdateam.ru.forpda.entity.remote.topics.TopicItem
import forpdateam.ru.forpda.entity.remote.topics.TopicsData
import forpdateam.ru.forpda.model.data.remote.api.favorites.FavoritesApi
import forpdateam.ru.forpda.model.repository.faviorites.FavoritesRepository
import forpdateam.ru.forpda.model.repository.forum.ForumRepository
import forpdateam.ru.forpda.model.repository.topics.TopicsRepository
import forpdateam.ru.forpda.presentation.ILinkHandler
import forpdateam.ru.forpda.presentation.Screen
import forpdateam.ru.forpda.presentation.TabRouter

/**
 * Created by radiationx on 03.01.18.
 */

@InjectViewState
class TopicsPresenter(
        private val topicsRepository: TopicsRepository,
        private val forumRepository: ForumRepository,
        private val favoritesRepository: FavoritesRepository,
        private val router: TabRouter,
        private val linkHandler: ILinkHandler
) : BasePresenter<TopicsView>() {

    var id = 0
    private var currentSt = 0
    var currentData: TopicsData? = null

    override fun onFirstViewAttach() {
        super.onFirstViewAttach()
        loadTopics()
    }

    fun loadTopics() {
        topicsRepository
                .getTopics(id, currentSt)
                .doOnTerminate { viewState.setRefreshing(true) }
                .doAfterTerminate { viewState.setRefreshing(false) }
                .subscribe({
                    viewState.showTopics(it)
                }, {
                    this.handleErrorRx(it)
                })
                .addToDisposable()
    }

    fun loadPage(st: Int) {
        currentSt = st
        loadTopics()
    }

    fun addForumToFavorite(forumId: Int, subType: String) {
        favoritesRepository
                .editFavorites(FavoritesApi.ACTION_ADD_FORUM, -1, forumId, subType)
                .subscribe({
                    viewState.onAddToFavorite(it)
                }, {
                    it.printStackTrace()
                })
                .addToDisposable()
    }

    fun addTopicToFavorite(topicId: Int, subType: String) {
        favoritesRepository
                .editFavorites(FavoritesApi.ACTION_ADD, -1, topicId, subType)
                .subscribe({
                    viewState.onAddToFavorite(it)
                }, {
                    it.printStackTrace()
                })
                .addToDisposable()
    }

    fun markRead() {
        forumRepository
                .markRead(id)
                .subscribe({
                    viewState.onMarkRead()
                }, {
                    it.printStackTrace()
                })
                .addToDisposable()
    }

    fun openForum() {
        router.navigateTo(Screen.Forum().apply {
            forumId = id
        })
    }

    fun openSearch() {
        router.navigateTo(Screen.Search().apply {
            searchUrl = "https://4pda.ru/forum/index.php?act=search&source=all&forums%5B%5D=$id"
        })
    }

    fun openTopicForum() {
        currentData?.let {
            linkHandler.handle("https://4pda.ru/forum/index.php?showforum=${it.id}", router)
        }
    }

    fun onItemClick(item: TopicItem) {
        if (item.isAnnounce) {
            linkHandler.handle(item.announceUrl, router, mapOf(
                    Screen.ARG_TITLE to item.title
            ))
            return
        }
        if (item.isForum) {
            linkHandler.handle("https://4pda.ru/forum/index.php?showforum=${item.id}", router)
            return
        }
        linkHandler.handle("https://4pda.ru/forum/index.php?showtopic=${item.id}&view=getnewpost", router, mapOf(
                Screen.ARG_TITLE to item.title
        ))
    }

    fun onItemLongClick(item: TopicItem) {
        viewState.showItemDialogMenu(item)
    }
}
